internal/context/user_plane_information.go

```diff
@@ -177,6 +178,244 @@ func NewUserPlaneInformation(upTopology *factory.UserPlaneInformation) *UserPlan
+			continue
+		}
+		upNode := new(UPNode)
+		upNode.Type = UPNodeType(node.Type)
+		switch upNode.Type {
+		case UPNODE_UPF:
+			// ParseIp() always return 16 bytes
+			// so we can't use the length of return ip to separate IPv4 and IPv6
+			// This is just a work around
+			var ip net.IP
+			if net.ParseIP(node.NodeID).To4() == nil {
+				ip = net.ParseIP(node.NodeID)
+			} else {
+				ip = net.ParseIP(node.NodeID).To4()
+			}
+
+			switch len(ip) {
+			case net.IPv4len:
+				upNode.NodeID = pfcpType.NodeID{
+					NodeIdType: pfcpType.NodeIdTypeIpv4Address,
+					IP:         ip,
+				}
+			case net.IPv6len:
+				upNode.NodeID = pfcpType.NodeID{
+					NodeIdType: pfcpType.NodeIdTypeIpv6Address,
+					IP:         ip,
+				}
+			default:
+				upNode.NodeID = pfcpType.NodeID{
+					NodeIdType: pfcpType.NodeIdTypeFqdn,
+					FQDN:       node.NodeID,
+				}
+			}
+
+			upNode.UPF = NewUPF(&upNode.NodeID, node.InterfaceUpfInfoList)
+			snssaiInfos := make([]SnssaiUPFInfo, 0)
+			for _, snssaiInfoConfig := range node.SNssaiInfos {
+				snssaiInfo := SnssaiUPFInfo{
+					SNssai: SNssai{
+						Sst: snssaiInfoConfig.SNssai.Sst,
+						Sd:  snssaiInfoConfig.SNssai.Sd,
+					},
+					DnnList: make([]DnnUPFInfoItem, 0),
+				}
+
+				for _, dnnInfoConfig := range snssaiInfoConfig.DnnUpfInfoList {
+					ueIPPools := make([]*UeIPPool, 0)
+					for _, pool := range dnnInfoConfig.Pools {
+						ueIPPool := NewUEIPPool(&pool)
+						if ueIPPool == nil {
+							logger.InitLog.Fatalf("invalid pools value: %+v", pool)
+						} else {
+							ueIPPools = append(ueIPPools, ueIPPool)
+						}
+					}
+					snssaiInfo.DnnList = append(snssaiInfo.DnnList, DnnUPFInfoItem{
+						Dnn:             dnnInfoConfig.Dnn,
+						DnaiList:        dnnInfoConfig.DnaiList,
+						PduSessionTypes: dnnInfoConfig.PduSessionTypes,
+						UeIPPools:       ueIPPools,
+					})
+				}
+				snssaiInfos = append(snssaiInfos, snssaiInfo)
+			}
+			upNode.UPF.SNssaiInfos = snssaiInfos
+			upi.UPFs[name] = upNode
+		case UPNODE_AN:
+			upNode.ANIP = net.ParseIP(node.ANIP)
+			upi.AccessNetwork[name] = upNode
+		default:
+			logger.InitLog.Warningf("invalid UPNodeType: %s\n", upNode.Type)
+		}
+
+		upi.UPNodes[name] = upNode
+
+		ipStr := upNode.NodeID.ResolveNodeIdToIp().String()
+		upi.UPFIPToName[ipStr] = name
+
+		// AllocateUPFID
+		upfid := upNode.UPF.UUID()
+		upfip := upNode.NodeID.ResolveNodeIdToIp().String()
+		upi.UPFsID[name] = upfid
+		upi.UPFsIPtoID[upfip] = upfid
+	}
+
+	// overlap UE IP pool validation
+	allUEIPPools := []*UeIPPool{}
+	for _, upf := range upi.UPFs {
+		for _, snssaiInfo := range upf.UPF.SNssaiInfos {
+			for _, dnn := range snssaiInfo.DnnList {
+				allUEIPPools = append(allUEIPPools, dnn.UeIPPools...)
+			}
+		}
+	}
+	if isOverlap(allUEIPPools) {
+		logger.InitLog.Fatalf("overlap cidr value between UPFs")
+	}
+}
+
+func (upi *UserPlaneInformation) LinksFromConfiguration(upTopology *factory.UserPlaneInformation) {
+	for _, link := range upTopology.Links {
```

Comment:
check if links already exist in SMF

Reply:
@tim-ywliu , 
I suggest that this code logic will be the same as done 
[here](https://github.com/free5gc/smf/blob/main/internal/context/user_plane_information.go#L155) where it does not validate the links.

I also noticed that if SMF encounters an IP overlap then it [aborts](https://github.com/free5gc/smf/blob/main/internal/context/user_plane_information.go#L152). I use the same logic  [here](https://github.com/aviweit/smf/blob/upi_rest_api/internal/context/user_plane_information.go#L402) as well.

Perhaps new PR can be created to deal with link validation, and IP overlap handling?

Reply:
@tim-ywliu , I was thinking that it would be good for this PR to focus on adding REST API functionality; new PRs will be opened to fix other existing SMF execution paths (e.g. validation, enhancements, ...). I hope it's ok.